contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,220 @@
+
+ES6 modules is also currently not natively supported by NodeJS, which still uses CommonJS. This explains why a lot of NodeJS applications still use CommonJS.
+
+All in all, the best implementation depends on the nature of your project.
+
+## How to start?
+
+You can start with module pattern right away by refactoring segments of your code into different files and wrapping the code in IIFE.
+
+If you are planning to use CommonJS or ES6 modules, you can start by refactoring segments of your code, but you will also need to be familiar with transpiler and bundler such as Babel and Webpack.
+
+## Further Reading
+
+You can read more on JavaScript modules at following websites:
+
+- [JavaScript Modules: A Beginner’s Guide](https://medium.freecodecamp.org/javascript-modules-a-beginner-s-guide-783f7d7a5fcc) <br>
+This article gives a broad explanation of JavaScript modules, explaining different kinds of module system in JavaScript, yet easy to follow.
+- [JavaScript Modules: From IIFEs to CommonJS to ES6 Modules](https://tylermcginnis.com/javascript-modules-iifes-commonjs-esmodules/) <br>
+This article truly focuses on IIFE, CommonJS, and ES6 Modules, giving a very comprehensive usage example.
+- [Learn the basics of the JavaScript module system and build your own library](https://medium.freecodecamp.org/anatomy-of-js-module-systems-and-building-libraries-fadcd8dbd0e) <br>
+While the aim of the article may not be the same with the aim of this book article, this article gives a very comprehesive comparison between CommonJS and ES6 Modules.
+
+Additional resources for related concepts (including all the hyperlinks above):
+
+1. Software engineering principles <br>
+[Software Engineering for Self-Directed Learners: SOLID Principles](http://se-education.org/se-book/principles/solidPrinciples/index.html) <br>
+[Software Engineering for Self-Directed Learners: Don't Repeat Yourself (DRY)](http://se-education.org/se-book/principles/dryPrinciple/index.html)
+
+2. JavaScript concepts <br>
+[Demystifying JavaScript Variable Scope and Hoisting](https://www.sitepoint.com/demystifying-javascript-variable-scope-hoisting/) <br>
+[IIFE - MDN Web Docs Glossary](https://developer.mozilla.org/en-US/docs/Glossary/IIFE)
+
+3. JavaScript development <br>
+[How JavaScript bundlers work](https://medium.com/@gimenete/how-javascript-bundlers-work-1fc0d0caf2da) <br>
+[JavaScript Transpilers: What They Are & Why We Need Them](https://scotch.io/tutorials/javascript-transpilers-what-they-are-why-we-need-them) <br>
+[What is tree shaking? 🌲](https://medium.com/@netxm/what-is-tree-shaking-de7c6be5cadd)
+
+4. Tools <br>
+[Bundler: Webpack](https://webpack.js.org/) <br>
+[Transpiler: Babel](https://babeljs.io/)
```

Comment:
After trimming the article as suggested above, you can remove irrelevant links too.

Reply:
Which suggestion?